sky/engine/platform/fonts/apple/FontCacheIOS.mm

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import <Foundation/Foundation.h>
+#import <CoreFoundation/CoreFoundation.h>
+#import <CoreText/CoreText.h>
+#import <UIKit/UIKit.h>
+
+#include "sky/engine/platform/fonts/FontCache.h"
+#include "sky/engine/platform/fonts/FontCache.h"
+#include "base/mac/scoped_nsobject.h"
+#include "base/lazy_instance.h"
+#include "base/macros.h"
+#include "base/logging.h"
+
+namespace blink {
+
+template <class T>
+class CFRef {
+ public:
+  CFRef() : _ref(nullptr) {}
+
+  CFRef(T t) : _ref(t) {}
+
+  CFRef(const CFRef&) = delete;
+
+  void operator=(const CFRef&) = delete;
+
+  CFRef(CFRef&& o) : _ref(o._ref) { o._ref = nullptr; }
+
+  void operator=(CFRef&& o) {
+    _ref = o._ref;
+    o._ref = nullptr;
+  }
+
+  operator bool() const { return _ref != nullptr; }
+
+  T get() const { return _ref; }
+
+  ~CFRef() {
+    if (_ref != nullptr) {
+      CFRelease(reinterpret_cast<CFTypeRef>(_ref));
+    }
+  }
+
+ private:
+  T _ref;
+};
+
+class FontFallbackSelector {
+ public:
+  FontFallbackSelector() {
+    UIFontDescriptor* ui_desc =
+        [UIFont systemFontOfSize:[UIFont systemFontSize]].fontDescriptor;
```

Comment:
Do we need to set up an Autorelease pool or Obj-C exception handlers?  It's been a long time since I've dealt with callouts from c++ to Obj-C from blink we used to have macros for such.